include/nektar_interface/projection/algorithm_types.hpp

```diff
@@ -0,0 +1,269 @@
+      return queue.submit([&](sycl::handler &cgh) {
+        cgh.parallel_for<>(range, [=](sycl::id<1> idx) {
+          auto cellx = data.cell_ids[idx];
+          auto npart = data.par_per_cell[cellx];
+
+          if (npart == 0)
+            return;
+
+          auto cell_dof = &data.dofs[data.dof_offsets[cellx]];
+          for (int part = 0; part < npart; ++part) {
+            auto xi0 = data.positions[cellx][0][part];
+            auto xi1 = data.positions[cellx][1][part];
+            T eta0, eta1;
+            Shape::template loc_coord_to_loc_collapsed<T>(xi0, xi1, eta0, eta1);
+            auto qoi = data.input[cellx][componant][part];
+            Shape::template project_one_particle<nmode, T, alpha, beta>(
+                eta0, eta1, qoi, cell_dof);
+          }
+        });
+      });
+    } else {
+      return queue.submit([&](sycl::handler &cgh) {
+        cgh.parallel_for<>(range, [=](sycl::id<1> idx) {
+          auto cellx = data.cell_ids[idx];
+          auto npart = data.par_per_cell[cellx];
+
+          if (npart == 0)
+            return;
+
+          auto cell_dof = &data.dofs[data.dof_offsets[cellx]];
+          for (int part = 0; part < npart; ++part) {
+            auto xi0 = data.positions[cellx][0][part];
+            auto xi1 = data.positions[cellx][1][part];
+            auto xi2 = data.positions[cellx][2][part];
+            T eta0, eta1, eta2;
+            Shape::template loc_coord_to_loc_collapsed<T>(xi0, xi1, xi2, eta0,
+                                                          eta1, eta2);
+            auto qoi = data.input[cellx][componant][part];
+            Shape::template project_one_particle<nmode, T, alpha, beta>(
+                eta0, eta1, eta2, qoi, cell_dof);
+          }
+        });
+      });
+    }
+  }
+};
+
+// Untested on Microsoft
+// c++20 has a solution for this std::source_location::function_name
+#ifdef _MSC_VER
+#define PRETTY_FUNCTION __FUNCSIG__
+#else
+#define PRETTY_FUNCTION __PRETTY_FUNCTION__
+#endif
+
+struct ThreadPerDof {
+
+  template <int nmode, typename T, int alpha, int beta, typename Shape>
+  std::optional<sycl::event> static inline project(DeviceData<T> &data,
+                                                       int componant,
+                                                       sycl::queue &queue) {
+
+    // TODO: What to do when local_size comes back as zero
+    // 1. Theoretically can use fewer than the warp size for the block size
+    //    just be slow
+    // 2. Crash always
+    // 3. Crash in debug/warn in release <--- doing this for now, crashing will
+    //    compilcate the benchmarking in release
+    auto local_size = Private::calc_max_local_size<nmode, T, Shape>(
+        queue, Constants::preferred_block_size);
+    if (local_size == 0) {
+      fprintf(stderr,
+              "%s: This kernel uses too much local(shared) memory for your "
+              "device\n",
+              PRETTY_FUNCTION);
+      // Crash in debug version
```

Comment:
This needs to be fatal in release as well.

Reply:
I don't crash there because it messes up the benchmarks and they are useful as is because they will just tell you what can't be run on whatever hardware you run  e.g on my laptop GPU
```
static std::optional<sycl::event> NESO::Project::ThreadPerDof::project(DeviceData<T> &, int, sycl::queue &) [nmode = 8, T = double, alpha = 1, beta = 1, Shape = NESO::Project::ePyramid<ThreadPerDof>]: This kernel uses too much local(shared) memory for your device
bm_project<8, double, ePyramid<ThreadPerDof>>   ERROR OCCURRED: 'Projection failed'
```
I'll check the option in the project_inner function and crash there